_posts/2020-10-09-from-sprockets-to-webpacker.markdown

```diff
@@ -0,0 +1,311 @@
+---
+layout: post
+title: "From Sprockets to Webpacker"
+date: 2020-10-09 11:00:00
+reviewed: 2020-10-09 11:00:00
+categories: ["rails", "webpack"]
+author: "arielj"
+---
+
+Back in 2011, [Rails 3.1 introduced The Assets Pipeline](https://guides.rubyonrails.org/3_1_release_notes.html#assets-pipeline) feature using the [Sprockets gem](https://github.com/rails/sprockets). This made it really easy to handle assets (images, fonts, JavaScript, CSS and more), solving many of the issues that developers had to face everyday.
+
+In 2012, [Webpack](https://webpack.js.org) was released solving the same issues in a different way and in time became the most used solution to handle assets. And since Rails 6, Webpack became the default solution to handle JavaScript assets using the [Webpacker gem](https://github.com/rails/webpacker) on new Rails applications.
+
+In this article I'll explain the steps we took to migrate to Webpack to handle our JavaScript assets, and how we added support as a node module in [our styleguide](https://github.com/fastruby/styleguide).
+
+<!--more-->
+
+## Main Differences
+
+There are a few key differences on how Sprockets and Webpack works:
+- Sprockets adds the content of required files into a one big file with all the code running globally and sequentially, Webpack creates a bundle that isolates each module following ES6 modules specs.
+- Sprockets uses `app/assets`, `vendor/assets` (also from gems) and `node_modules` folders to look for assets, while Webpacker uses the `app/javascript` and the `node_modules` folder.
+- Since the `// require ...` syntax is specific for Sprockets, it's not recognized by linters, while the `import` syntax of Webpack is the ES6 syntax so more tools understand that.
+
+## The Plan
+
+### FastRuby.io Styleguide
+
+We use our styleguide for the base style of our projects, but it works either as a Ruby gem or as static files. In order to use it with Webpack, we need to add support to be used as a node module (we'll use Yarn since it's the default node modules manager for Rails applications -like bundler for gems management-).
+
+### Webpack Only for JavaScript
+
+While it is possible to use Webpack for all kind of assets (images, styles, fonts, etc). We are only going to use it for JavaScript since it's the current Rails standard. This may change in the future if Webpack gets better and faster for those types [according to DHH](https://discuss.rubyonrails.org/t/sprockets-abandonment/74371/41).
+
+### 3rd Party Assets in Gems
+
+Some assets are added by gems, but, similar to the styleguide, it's easier to integrate them with Webpack if we use an equivalent Yarn package.
```

Comment:
@arielj Could you add an example? It might make it clearer to understand. 

Are you talking about the difference in including 3rd party libraries? For instance: If you use sprockets, then you would use something like the bootstrap-rails gem, but if you are using webpacker then you need to find the right npm package, is that it?

Reply:
Yes, that's what I meant, I'll add `@rails/ujs` as an example since I used that package during the actual steps.